### PR TITLE
CMaize Startup and Project Class

### PR DESCRIPTION
**Pull request by ryanmrichard**
**Is this pull request associated with an issue(s)?**
No

**Description**
This PR writes up some notes about what happens when CMaize is included by the user and describes the design of the ``CMaizeProject`` class (an instance of which is created when CMaize is loaded).

**TODOs**
- [x] Document ``CMaizeProject``.

### CODE REVIEW

**Comment by ryanmrichard** [comment]
@zachcran I'm looking at the `CMaizeProject` class. Do we track which targets are from dependencies and which are part of the project? In particular, I assume that in general the `build_targets` field will contain a mix of targets that may have either come from a dependency or the project? I also assume that the `installed_targets` field will only contain dependency targets (because I don't know how you have an installed version of the project you're building already).

**Comment by zachcran** [comment]
>Do we track which targets are from dependencies and which are part of the project?

I don't think we differentiate between targets generated from dependencies and targets generated from projects. `installed_targets` will likely only have dependency targets for the same reasons you gave, but `build_targets` can have both project and dependency targets in it.

**Comment by ryanmrichard** [comment]
I think this is r2g. I ultimately decided to combine the CMaizeProject and PackageSpecification components into one component (which I just refer to as the CMaizeProject component). This is how the source is laid out anyways. If we stuck with the previous design the CMaizeProject component would just be a single class (PackageSpecification would have been too, though I could see that evolving). This PR did result in some TODOs in #123.

File: docs/src/developer/design/assets/cmaize_project.png
Comment (zachcran) [inline docs/src/developer/design/assets/cmaize_project.png:1]: For the PackageSpecification note, it took me a few reads to understand "Input consumers of package provide to a PackageManager" here. Maybe "Provided to a PackageManager to get back the installed version of the package" would be clearer, if that still captures the message you are looking for.

In the PackageSpecification note, you also say "installed version of this project", which I think should be "installed version of this package".

File: docs/src/developer/design/cmaize_project.rst [hunk 1]
```diff
@@ -18,4 +18,180 @@
...
+easily see the pieces of the :term:`project`. For example, at the time of
+writing there is not a simple CMake function which returns a list of a
+project's :term:`dependencies <dependency>`. Without a mechanism to query the
+properties of a project, CMake relies on the user to resupply the same
+information to a number of functions.
+
+One of the main considerations for writing CMaize is
+:ref:`minimizing the redundancy<minimize_redundancy>` found in typical
+CMake-based :term:`build systems <build system>`. To this end, the CMaizeProject
+component is designed to collect the the project's properties into a single
```
Comment (zachcran) [inline docs/src/developer/design/cmaize_project.rst:51]:
```suggestion
component is designed to collect the project's properties into a single
```

File: docs/src/developer/design/cmaize_project.rst [hunk 2]
```diff
@@ -18,4 +18,180 @@
...
+to represent a project which uses CMaize as its build system. Together the
+:ref:`cp_building` and the :ref:`cp_packaging` considerations establish the
+state the ``CMaizeProject`` object must possess. The :ref:`cp_recursive`
+consideration leads to the realization that the current project (in package
+form) will in general be a dependency of one or more downstream projects. When
+the downstream project needs to find the current
+project, the downstream project will need the ability to uniquely distinguish
+among package iterations (instances of a package that differ in configuration
+and/or version). This requires tagging the package with some subset of a
+``CMaizeProject`` object's state and  to this end we introduce
```
Comment (zachcran) [inline docs/src/developer/design/cmaize_project.rst:151]:
```suggestion
``CMaizeProject`` object's state and to this end we introduce
```

**Review by zachcran — CHANGES_REQUESTED** [review]

**Pull request merged**